src/ocrd/processor/base.py

```diff
@@ -575,67 +605,66 @@ def zip_input_files(self, require_first=True, mimetype=None, on_error='skip'):
                     pageId=self.page_id, fileGrp=ifg, mimetype=mimetype),
                                 # sort by MIME type so PAGE comes before images
                                 key=lambda file_: file_.mimetype)
-            # Warn if no files found but pageId was specified because that
-            # might be because of invalid page_id (range)
-            if self.page_id and not files_:
-                msg = (f"Could not find any files for --page-id {self.page_id} - "
-                       f"compare '{self.page_id}' with the output of 'orcd workspace list-page'.")
-                if on_error == 'abort':
-                    raise ValueError(msg)
-                LOG.warning(msg)
             for file_ in files_:
                 if not file_.pageId:
+                    # ignore document-global files
                     continue
                 ift = pages.setdefault(file_.pageId, [None]*len(ifgs))
                 if ift[i]:
-                    LOG.debug("another file %s for page %s in input file group %s", file_.ID, file_.pageId, ifg)
+                    LOG.debug(f"another file {file_.ID} for page {file_.pageId} in input file group {ifg}")
                     # fileGrp has multiple files for this page ID
                     if mimetype:
                         # filter was active, this must not happen
+                        LOG.warning(f"added file {file_.ID} for page {file_.pageId} in input file group {ifg} "
+                                    f"conflicts with file {ift[i].ID} of same MIME type {mimetype} - on_error={on_error}")
                         if on_error == 'skip':
                             ift[i] = None
                         elif on_error == 'first':
                             pass # keep first match
                         elif on_error == 'last':
                             ift[i] = file_
                         elif on_error == 'abort':
-                            raise ValueError(
-                                "Multiple '%s' matches for page '%s' in fileGrp '%s'." % (
-                                    mimetype, file_.pageId, ifg))
+                            raise NonUniqueInputFile(ifg, file_.pageId, mimetype)
                         else:
                             raise Exception("Unknown 'on_error' strategy '%s'" % on_error)
```

Comment:
This should become a separate method. I see the same if checks are repeated below. The constant strings for the `on_error` strategy could also become an Enum class. The Enum class would also help with the validator and parser in the config.

Reply:
> This should become a separate method. I see the same if checks are repeated below.

We used to have two discerning error messages here. (In one case, a specific MIME type was requested, like PAGE-only or image-only or JSON or whatever, but that was not unique. In another, no MIME type was requested, so the PAGE+images logic should apply but did not establish uniqueness via some PAGE file.)

How would you like to see this refactored into a function, exactly?

> The constant strings for the `on_error` strategy could also become an Enum class. 

Agreed.

> The Enum class would also help with the validator and parser in the config.

Remember that this concrete setting should not be decided by the user. The config has a different granularity level: what should happen to input files missing (or skipped because of being non-unique).

Reply:
> How would you like to see this refactored into a function, exactly?

Something like:

```python
def handle_errors(ift, file_, ifg: str, on_error: str):
  if on_error == 'skip':
      ift[i] = None
  elif on_error == 'first':
      pass # keep first match
  elif on_error == 'last':
      ift[i] = file_
  elif on_error == 'abort':
      raise NonUniqueInputFile(ifg, file_.pageId, None)
  else:
      raise Exception("Unknown 'on_error' strategy '%s'" % on_error)
```

The only difference I see is the warning logs content which could be preserved outside the method. If different things happened, say on skip and abort, then we no longer have the method option. Anyway, nothing important for the current phase of v3.

Reply:
Ok, sure, we can pass in the differentiating part of the message for NonUniqueInputFile.

But this is problematic:

> ```python
> if on_error == 'skip':
>       return None
>   elif on_error == 'first':
>       return None
> ```

How would you call that function?

With skip, we should set the input file to None. With first, we should keep the previous OcrdFile instance.

Reply:
Right, the return version would not work and the other version requires passing some arguments for just checking the on_error state. Let's leave it as it is.